sandbox/host/src/sandbox/wasmi_backend.rs

```diff
@@ -18,393 +18,544 @@
+    let mut linker = Linker::new(store.engine());
+
+    for import in module.imports() {
+        let module = import.module().to_string();
+        let name = import.name().to_string();
+
+        match import.ty() {
+            ExternType::Global(_) | ExternType::Table(_) => {}
+            ExternType::Memory(_mem_ty) => {
+                let memory = guest_env
+                    .imports
+                    .memory_by_name(&module, &name)
+                    .ok_or(InstantiationError::ModuleDecoding)?;
+
+                let wasmi_memory = memory.as_wasmi().expect(
+                    "memory is created by wasmi; \
+                    exported by the same module and backend; \
+                    thus the operation can't fail; \
+                    qed",
+                );
+
+                linker
+                    .define(&module, &name, wasmi_memory.memory)
+                    .map_err(|_| InstantiationError::EnvironmentDefinitionCorrupted)?;
+            }
+            ExternType::Func(func_ty) => {
+                let guest_func_index = guest_env
+                    .imports
+                    .func_by_name(import.module(), import.name());
+
+                let guest_func_index = if let Some(index) = guest_func_index {
+                    index
+                } else {
+                    // Missing import (should we abort here?)
+                    continue;
+                };
+
+                let supervisor_func_index = guest_env
+                    .guest_to_supervisor_mapping
+                    .func_by_guest_index(guest_func_index)
+                    .ok_or(InstantiationError::ModuleDecoding)?;
+
+                let function = match version {
+                    Instantiate::Version1 => {
+                        dispatch_function(supervisor_func_index, &mut store, func_ty)
+                    }
+                    Instantiate::Version2 => {
+                        dispatch_function_v2(supervisor_func_index, &mut store, func_ty)
+                    }
+                };
 
-    with_guest_externals(&sandbox_instance, |guest_externals| {
-        SupervisorContextStore::using(supervisor_context, || {
-            wasmi_instance
-                .run_start(guest_externals)
-                .map_err(|_| InstantiationError::StartTrapped)
-        })
+                linker
+                    .define(&module, &name, function)
+                    .map_err(|_| InstantiationError::ModuleDecoding)?;
+            }
+        }
+    }
+
+    let instance_pre = linker.instantiate(&mut *store, &module).map_err(|error| {
+        log::trace!("Failed to call wasmi instantiate: {error:?}");
+        InstantiationError::Instantiation
+    })?;
+    let instance = instance_pre.start(&mut *store).map_err(|error| {
+        log::trace!("Failed to call wasmi start: {error:?}");
+        InstantiationError::StartTrapped
     })?;
```

Comment:
This is where `_supervisor_context: &mut dyn SupervisorContext` must be used

Reply:
It won't be used because ATM, we can't have start functions in programs, but I can add it for symmetry with the Wasmer backend implementation.

Reply:
We can't for now but we want in the future: https://github.com/gear-tech/gear/issues/938

Reply:
Fixed